[kernel 3b/n] Decouple `{Dump,Load}Mempool` from `ArgsManager`

dongcarl opened the pull request:
This is part of the `libbitcoinkernel` project: #24303, https://github.com/bitcoin/bitcoin/projects/18

-----

This PR moves `{Dump,Load}Mempool` into its own `kernel/mempool_persist` module and introduces `ArgsManager` `node::` helpers in `node/mempool_persist_args`to remove the scattered calls to `GetBoolArg("-persistmempool", DEFAULT_PERSIST_MEMPOOL)`.

More context can be gleaned from the commit messages.

-----

One thing I was reflecting on as I wrote this was that in the long run, I think we should probably invert the validation <-> mempool relationship. Instead of mempool not depending on validation, it might make more sense to have validation not depend on mempool. Not super urgent since `libbitcoinkernel` will include both validation and mempool, but perhaps something for the future.

dongcarl committed:
DumpMempool: Use std::chrono instead of weird int64_t arthmetics

This makes it so that DumpMempool doesn't depend on MICRO anymore

dongcarl committed:
test/fuzz: Invoke LoadMempool via CChainState

Not only does this increase coverage, it is also more correct in that
when ::LoadMempool is called with a mempool and chainstate, it calls
AcceptToMemoryPool with just the chainstate.

AcceptToMemoryPool will then act on the chainstate's mempool via
CChainState::GetMempool, which may be different from the mempool
originally passed to ::LoadMempool. (In this fuzz test's case, it
definitely is different)

Also, move DummyChainstate to its own file since it's now used by the
validation_load_mempool fuzz test to replace CChainState's m_mempool.

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -1027,6 +1034,161 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+        unbroadcast_txids = GetUnbroadcastTxs();
+    }
+
+    auto mid = GetSystemTime<std::chrono::microseconds>();
```
When touching this line, it could make sense to make it a behaviour change and use steady clock, see for example #25456.

dongcarl commented:
@MarcoFalke 

Following up on https://github.com/bitcoin/bitcoin/pull/25290#discussion_r908811064, wondering what clock I should be using there? SteadyClock? NodeClock?

Also, could you sanity check my clock switch in 0ea35ed965725341428a657c4fb29c518969af60? Thanks!

MarcoFalke commented:
Jup, steady clock is fine, given that previously system clock was used and the result is only used for logging a duration (time difference).

`Ticks<SecondsDouble>` should also work?

dongcarl commented:
@MarcoFalke Okay, I'll use `Ticks<SecondsDouble>` and steady clock for DumpMempool, what about for LoadMemPool you mentioned here? https://github.com/bitcoin/bitcoin/pull/25290#discussion_r908811064

MarcoFalke commented:
The timestamps for mempool entries are mockable (GetTime), thus they are (supposed to be) using NodeClock.

dongcarl committed:
mempool: Use NodeClock+friends for LoadMempool

dongcarl commented:
Push 05ff88acdfc1b777e7d6e53a2975e18ead6f35b1:
- Use type-safe time types

-----

I think this PR is ready for review!

MarcoFalke commented on src/kernel/mempool_options.h:
```diff
@@ -32,6 +34,8 @@ struct MemPoolOptions {
     std::chrono::seconds expiry{std::chrono::hours{DEFAULT_MEMPOOL_EXPIRY_HOURS}};
     MemPoolLimits limits{};
+    /* An empty path turns off mempool persistence */
+    fs::path persistence_path = {};
```
nit in 5b426d71aab1d2c7a3f7b44043e98a0e9b821518:

No need for ` = `

MarcoFalke commented on src/mempool_args.h:
```diff
@@ -10,6 +10,9 @@ namespace kernel {
 };
 
+/** Default for -persistmempool */
+static constexpr bool DEFAULT_PERSIST_MEMPOOL = true;
```
nit in 5b426d71aab1d2c7a3f7b44043e98a0e9b821518:

Could use `{}`.

dongcarl replied:
Fixed as of 930f023ea9

MarcoFalke commented on src/util/time.h:
```diff
@@ -71,6 +71,9 @@ int64_t GetTimeMillis();
 int64_t GetTimeMicros();
 
+template <typename T>
+T GetSystemTime();
```
276cfd2f5fd48d0556ccc4a83d350c0e9d45cc99: why is this file changed?

dongcarl replied:
Sorry, it was leftover... Fixed as of 6612412dfc

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4773,8 +4777,11 @@ bool DumpMempool(const CTxMemPool& pool, FopenFn mockable_fopen_function, bool s
+
+        auto copy_secs = Ticks<SecondsDouble>(mid - start);
+        auto dump_secs = Ticks<SecondsDouble>(last - mid);
+        LogPrintf("Dumped mempool: %gs to copy, %gs to dump\n", copy_secs, dump_secs);
```
nit: Could keep inlined to not evaluate the args if the log is disabled ?

dongcarl replied:
Done as of 6612412dfc

MarcoFalke commented on src/validation.cpp:
```diff
@@ -47,14 +47,12 @@
 #include <warnings.h>
 
 #include <algorithm>
-#include <chrono>
```
35a04e0707478f18a5c12e8979a823e119d9: Are you sure that nothing else uses those two headers?

dongcarl replied:
Ah my bad, reinstated as of a19cc42ca2

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -1150,7 +1149,7 @@ bool CTxMemPool::_LoadMempool(std::function<bool(const CTransactionRef&, int64_t
                 PrioritiseTransaction(tx->GetHash(), amountdelta);
             }
-            if (nTime > nNow - nExpiryTimeout) {
+            if (nTime > TicksSinceEpoch<Seconds<int64_t>>(now - m_expiry)) {
```
05ff88acdfc1b777e7d6e53a2975e18ead6f35b1:

What do you think about `if (std::chrono::seconds{nTime}  > now - m_expiry) {` ?

dongcarl replied:
I tried it, and I think the only thing that'll compile is:

```cpp
            if (NodeSeconds{std::chrono::seconds{nTime}} > now - m_expiry) {
```

Which doesn't seem better than what we have now

MarcoFalke replied:
I see. Is there any reason why you need `Seconds<int64_t>`? `std::chrono::seconds` should already be of the same width.

dongcarl replied:
From reading the docs, the representation for `std::chrono::seconds` seems to only need to be "A signed integral type of at least 35 bits", so there might be platforms/stdlib's that don't use `int64_t` and I thought it'd be better to be explicit about it.

MarcoFalke replied:
Is there anything that requires this code part to be explicit about it? Does it protect against something?

dongcarl replied:
It doesn't really, happy to change it to `std::chrono::seconds` if you want me to!

MarcoFalke replied:
If there is no reason to change util/time.h, then I think the preference is to not change it.

dongcarl replied:
Agreed! Done in ddf4920df58bc6f84297abc9b31e5830237ea514

maflcko reviewed (approved):
left some nits (feel free to ignore)

MarcoFalke commented on src/Makefile.test_fuzz.include:
```diff
@@ -10,7 +10,8 @@ EXTRA_LIBRARIES += \
     test/fuzz/FuzzedDataProvider.h \
-    test/fuzz/util.h
+    test/fuzz/util.h \
+    test/fuzz/mempool_utils.h
```
nit: keep sorted with `m` before `u`?

dongcarl replied:
Fixed as of 2f8699a9423c048b1b36a13ea020e687dae245f2

glozow reviewed (commented):
Concept + Approach ACK

theuni commented on src/txmempool.cpp:
```diff
@@ -1105,6 +1106,96 @@ bool CTxMemPool::DumpMempool(FopenFn mockable_fopen_function, bool skip_file_com
+void CTxMemPool::LoadMempool(std::function<bool(const CTransactionRef&, int64_t)> simple_atmp, FopenFn mockable_fopen_function)
+{
+    _LoadMempool(simple_atmp, mockable_fopen_function);
+    SetIsLoaded(!ShutdownRequested());
```
This is really confusing. `_LoadMempool` may return false if `ShutdownRequested()`, but that's ignored here and it's checked again.

It's also not clear _why_ ShutdownRequested() may have been called. Might it be coming from some low-level operation stemming from `_LoadMempool` itself? Or is it `simple_atmp` we're worried about?

I see that you described the confusing semantics in the commit message, but I'm not convinced that it's ok to double-down on the wonky behavior before improving it some :(

dongcarl replied:
Followup entry added in umbrella issue #24303

theuni commented on src/txmempool.cpp:
```diff
@@ -1127,7 +1126,7 @@ bool CTxMemPool::_LoadMempool(std::function<bool(const CTransactionRef&, int64_t
     int64_t already_there = 0;
     int64_t unbroadcast = 0;
-    int64_t nNow = GetTime();
+    auto now = NodeClock::now();
```
Any reason not to use a steady clock here since it's an interval?

dongcarl replied:
Closing, feel free to reopen if not resolved!

theuni commented on src/validation.cpp:
```diff
@@ -3860,13 +3860,13 @@ void PruneBlockFilesManual(CChainState& active_chainstate, int nManualPruneHeigh
-    }
-    m_mempool->SetIsLoaded(!ShutdownRequested());
+    auto simple_atmp = [&](const CTransactionRef& tx, int64_t accept_time) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+        return AcceptToMemoryPool(*this, tx, accept_time, /*bypass_limits=*/false, /*test_accept=*/false).m_result_type == MempoolAcceptResult::ResultType::VALID;
```
Do we need all of `AcceptToMemoryPool` here? Without digging too much, I should think we might only need a subset, potentially skipping some of the uncaching/flushing logic?

dongcarl replied:
I'm not sure, and I don't know enough about this logic to confidently say. I'd prefer to preserve as much existing behavior as possible for an important module like mempool, which is what I think this does.

Okay to leave as a followup for someone with more knowledge of this logic?

dongcarl replied:
Followup entry added in umbrella issue #24303

theuni reviewed (commented):
Concept ACK and light review ACK.

I left a few comments/complaints about cleaning up some of the call/return semantics. I'd feel more comfortable with these changes if they didn't preserve some of our current weirdness, but  I don't have a ton of concrete suggestions without really digging in.

dongcarl commented on src/mempool_args.cpp:
```diff
@@ -33,5 +33,12 @@ void ApplyArgsManOptions(const ArgsManager& argsman, MemPoolOptions& mempool_opt
 
+
+    if (argsman.GetBoolArg("-persistmempool", DEFAULT_PERSIST_MEMPOOL)) {
+        if (mempool_opts.persistence_path.empty()) mempool_opts.persistence_path = argsman.GetDataDirNet() / "mempool.dat";
```
Mostly because the other options specifies a value that goes into `mempool_opts`, whereas `-persistmempool` is a boolean option.

Note that we use an empty `persistence_path` to indicate that we don't want to persist the mempool

So the logic goes:

- If we want to persist the mempool, and there's no existing mempool path set, set `persistence_path` to `<datadir>/ mempool.dat`
- If we don't want to persist the mempool, set `persistence_path` to an empty path

dongcarl commented on src/txmempool.cpp:
```diff
@@ -1036,6 +1039,73 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    std::vector<TxMempoolInfo> vinfo;
+    std::set<uint256> unbroadcast_txids;
+
+    static Mutex dump_mutex;
```
As a non-`static` or as a `static`? I think as a `static` it may prevent cases where 2 mempools are created with the same persistence path and try to dump at the same time. However, I think that's perhaps somewhat nonsensical anyway. I don't care either way.

theuni replied:
In that case I'd expect a lock at the filesystem level as opposed to a mutex. But sure, it's not at all a priority, I was just curious because I assumed the reason it was static before is because it wasn't well encapsulated.

dongcarl replied:
Followup entry added in umbrella issue #24303

MarcoFalke replied:
This is used consistently in the codebase, see also `DumpBanlist`.

theuni commented:
> For larger things like [simplifying `AcceptToMemoryPool`](https://github.com/bitcoin/bitcoin/pull/25487#discussion_r916059908) and [changing `_LoadMempool`'s semantics](https://github.com/bitcoin/bitcoin/pull/25487#discussion_r916044902), I'm going to add extensive documentation of what I've found, and leave it for followups. These seem like delicate operations which can either slow down libbitcoinkernel progress or introduce subtle bugs. I think they're better off as standalone efforts.

I think that's perfectly reasonable as long as nothing here gets more convoluted (arguably the atmp callback does that a little, but that's reasonable enough), and the current status quo is fully understood and documented. +1 for more docs, thanks for that :)

dongcarl committed:
scripted-diff: Rename m_is_loaded -> m_load_tried

m_is_loaded/IsLoaded() doesn't actually indicate whether or not the
mempool was successfully, loaded, but rather if a load has been
attempted and did not result in a catastrophic ShutdownRequested.

-BEGIN VERIFY SCRIPT-
find_regex="\bm_is_loaded\b" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@m_load_tried@g"

find_regex="\bIsLoaded\b" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@GetLoadTried@g"

find_regex="\bSetIsLoaded\b" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@SetLoadTried@g"
-END VERIFY SCRIPT-

dongcarl committed:
test/mempool_persist: Test manual savemempool when -persistmempool=0

dongcarl commented:
Push ddf4920df58bc6f84297abc9b31e5830237ea514 -> 8bf70da49cb0f79c7c09d126a37942a76fad7093
- Add a functional test for `savemempool` when `-persistmempool=0`
- IWYU `src/mempool_args.cpp`
- Made `savemempool` work when `-persistmempool=0`
- Rename `LoadMempool` -> `TryLoadMempool`
- Rename `m_is_loaded` -> `m_load_tried`
- Add a commit that makes `mapNextTx` private

dongcarl commented on src/txmempool.cpp:
```diff
@@ -1105,6 +1106,96 @@ bool CTxMemPool::DumpMempool(FopenFn mockable_fopen_function, bool skip_file_com
     return true;
 }
 
+void CTxMemPool::LoadMempool(std::function<bool(const CTransactionRef&, int64_t)> simple_atmp, FopenFn mockable_fopen_function)
```
Fixed!

MarcoFalke commented:
From CI:

`FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\test_runner_₿_🏃_20220708_004922\\mempool_persist_175\\node0\\regtest\\mempool.dat' -> 'C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\test_runner_₿_🏃_20220708_004922\\mempool_persist_175\\node1\\regtest\\mempool.dat'`

ariard commented on src/txmempool.h:
```diff
@@ -735,10 +735,10 @@ class CTxMemPool
 
     /** @returns true if the mempool is fully loaded */
-    bool IsLoaded() const;
+    bool GetLoadTried() const;
```
comment could be updated to reflect it indicates load attempt not succes

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25487#discussion_r917203102

In commit "scripted-diff: Rename m_is_loaded -> m_load_tried" (ae6bf8ed30d5b8511f22d3a8f16112b33804944f)

> comment could be updated to reflect it indicates load attempt not succes

In case you do manually update the comment after the scripted diff, could also renamed the "bool loaded" argument below "bool load_tried"

dongcarl replied:
Fixed in 82111033f3

ariard commented on src/txmempool.h:
```diff
@@ -821,7 +818,15 @@ class CTxMemPool
+    /** Sets the current loaded state */
+    void SetLoadTried(bool loaded);
+
+    bool LoadMempool(std::function<bool(const CTransactionRef&, int64_t)> simple_atmp, FopenFn mockable_fopen_function = fsbridge::fopen);
```
can get a comment

dongcarl replied:
The newest version of this PR doesn't have `simple_atmp`, and the new `::LoadMempool` signature seem simple enough, feel free to reopen if still a problem!

ariard commented on src/validation.cpp:
```diff
@@ -3860,13 +3860,13 @@ void PruneBlockFilesManual(CChainState& active_chainstate, int nManualPruneHeigh
-        ::LoadMempool(*m_mempool, *this);
-    }
-    m_mempool->SetLoadTried(!ShutdownRequested());
+    auto simple_atmp = [&](const CTransactionRef& tx, int64_t accept_time) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
```
I guess `simple_atmp` in opposition to the more complex `AcceptPackage` / `AcceptMultipleTransactions` ? In that case `single_tx_atmp` might be more speaking, I don't know.

dongcarl replied:
No longer relevant as of 80b303ea1a065602cf3f84d45b04dce95bfb6952

ariard reviewed (commented):
Approach ACK

ryanofsky commented on src/txmempool.h:
```diff
@@ -573,6 +574,8 @@ class CTxMemPool
 
     const Limits m_limits;
 
+    const fs::path m_persistence_path{};
```
In commit "mempool: Introduce m_persistence_path member" (ea750b1374fa0a49212b1f280660be7b2f6b9b2f)

I suspect it would simplify the PR and the code overall not to add a `CTxMemPool::m_persistence_path` member or an `MemPoolOptions::persistence_path` option. You could avoid the noise and statefulness and the twisty `const fs::path dump_path{mempool.m_persistence_path.empty() ? MemPoolPersistencePath(args) : mempool.m_persistence_path};` logic by writing a simple helper functions:

```c++
bool MempoolPersist(const ArgsManager& args)
{
    return args.GetBoolArg("-persistmempool", DEFAULT_PERSIST_MEMPOOL));
}

fs::path MempoolPath(const ArgsManager& args)
{
    return args.GetDataDirNet() / "mempool.dat";
}
```

And calling them on as needed from validation and RPC code without adding extra options and state.

theuni replied:
Mmm, this would simplify the code in question, though I personally prefer the pattern of parsing these upfront to avoid having to involve ArgsManager for something that seems like a property of `CTxMemPool`.

dongcarl replied:
Implemented as of 80b303ea1a065602cf3f84d45b04dce95bfb6952

ryanofsky reviewed (commented):
A lot of the changes seem good here, but I think I found a bug in the dump function, and I would suggest not burdening the `CTxMemPool` class with this persistence functionality, and instead moving the dump/load functions to a separate `mempool_persist.cpp` file so mempool code can be more focused and maintainable.

theuni commented on src/txmempool.cpp:
```diff
@@ -1036,6 +1039,73 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+
+    {
+        LOCK(cs);
+        for (const auto &i : mapDeltas) {
```
+1, nice catch!

dongcarl replied:
As of 80b303ea1a065602cf3f84d45b04dce95bfb6952, I rebased over #25592 which should test for this, and the bug no longer exists with the new approach

theuni commented on src/txmempool.cpp:
```diff
@@ -1036,6 +1039,73 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
     return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 15 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + memusage::DynamicUsage(vTxHashes) + cachedInnerUsage;
 }
 
+bool CTxMemPool::DumpMempoolPath(const fs::path& dump_path, FopenFn mockable_fopen_function, bool skip_file_commit) const
```
I think I agree with this as well, though I'm not very familiar with this code. Since dump/load kinda straddles `CChainState` and `CTxMemPool` neither seem like a great home, so standalone makes sense to me.

dongcarl replied:
Done as of 80b303ea1a065602cf3f84d45b04dce95bfb6952

dongcarl committed:
DumpMempool: Pass in dump_path, stop using gArgs

Also introduce node::{ShouldPersistMempool,MempoolPath} helper functions
in node/mempool_persist_args.{h,cpp} which are used by non-kernel
DumpMempool callers to determine whether or not to automatically dump
the mempool and where to dump it to.

dongcarl committed:
LoadMempool: Pass in load_path, stop using gArgs

Also:
1. Have CChainState::LoadMempool and ::ThreadImport take in paths and
   pass it through untouched to LoadMempool.
2. Make LoadMempool exit early if the load_path is empty.
3. Adjust the call to ::ThreadImport in ::AppInitMain to correctly pass
   in an empty path if mempool persistence is disabled.

dongcarl committed:
Move {Load,Dump}Mempool to kernel namespace

Also:
1. Add the newly introduced kernel/mempool_persist.cpp to IWYU CI script
2. Add chrono mapping for iwyu

dongcarl committed:
Move FopenFn to fsbridge namespace

[META] In a future commit in this patchset, it will be used by more than
       just validation, and it needs to align with fopen anyway.

dongcarl commented:
Push 4a4102d0986d88f548ae3ea3d5bcca10a2e598a2 -> 80b303ea1a065602cf3f84d45b04dce95bfb6952
- Rework PR to move `{Load,Dump}Mempool` to their own `mempool_persist` module instead of moving them to be methods of `CTxMemPool`

dongcarl commented on test/functional/mempool_persist.py:
```diff
@@ -133,6 +133,15 @@ def run_test(self):
+        mempooldat1 = os.path.join(self.nodes[1].datadir, self.chain, 'mempool.dat')
+
+        self.log.debug("Force -persistmempool=0 node1 to savemempool to disk via RPC")
+        result1 = self.nodes[1].savemempool()
```
Done as of c84390b741ab7b61c9f702d8b447c8cadc1257c8

dongcarl committed:
Move DEFAULT_PERSIST_MEMPOOL out of libbitcoinkernel

It is no longer used by anything inside libbitcoinkernel, move it to
node/mempool_persist_args.h where it belongs.

dongcarl commented:
Push 80b303ea1a065602cf3f84d45b04dce95bfb6952 -> fda55102298c5a9a280173e5a4a8e30e5943a447
- Added commit 267cbd1bd40af4bf52a0905cafb303c8f2c2bfac

jamesob reviewed (approved):
ACK fda55102298c5a9a280173e5a4a8e30e5943a447 ([`jamesob/ackr/25487.1.dongcarl.kernel_3b_n_decouple_du`](https://github.com/jamesob/bitcoin/tree/ackr/25487.1.dongcarl.kernel_3b_n_decouple_du))

Very straightforward now after a few rounds of review. Looks good!

Read through commits locally and built all.

dongcarl commented on ci/test/06_script_b.sh:
```diff
@@ -41,6 +41,7 @@ if [ "${RUN_TIDY}" = "true" ]; then
   CI_EXEC "python3 ${DIR_IWYU}/include-what-you-use/iwyu_tool.py"\
           " src/compat"\
           " src/init"\
+          " src/kernel/mempool_persist.cpp"\
```
Unfortunately `iwyu_tool.py` doesn't accept header files as individual arguments, it does check the corresponding header file when a `.cpp` is specified tho.

dongcarl commented:
Push fda55102298c5a9a280173e5a4a8e30e5943a447 -> 6f018a1f908f264667b024e60bb2394fed48f05c
- Addressed James' reviews:
  - https://github.com/bitcoin/bitcoin/pull/25487#discussion_r920428471
  - https://github.com/bitcoin/bitcoin/pull/25487#discussion_r920431537)

dongcarl commented on src/node/mempool_persist_args.cpp:
```diff
@@ -0,0 +1,23 @@
+
+namespace node {
+
+bool MempoolPersist(const ArgsManager& argsman)
```
Fixed in 04f72e9020

dongcarl commented:
Push 6f018a1f908f264667b024e60bb2394fed48f05c -> 2480e6823044b1fe6ba8fc84ee4434437a8c9df0
- Fix me being a bit careless

jamesob reviewed (approved):
reACK 2480e6823044b1fe6ba8fc84ee4434437a8c9df0 ([`jamesob/ackr/25487.2.dongcarl.kernel_3b_n_decouple_du`](https://github.com/jamesob/bitcoin/tree/ackr/25487.2.dongcarl.kernel_3b_n_decouple_du))

https://github.com/bitcoin/bitcoin/pull/25487/commits/ae6bf8ed30d5b8511f22d3a8f16112b33804944f may be the prettiest scripted-diff I ever have seen.

ryanofsky reviewed (approved):
Code review ACK 2480e6823044b1fe6ba8fc84ee4434437a8c9df0. Very simple and clean changes

dongcarl committed:
mempool: Improve comments for [GS]etLoadTried

Also change the param name for SetLoadTried to load_tried.

dongcarl commented:
Push 2480e68230 -> 3188375d36
- Added more comments, addressing: 
  - https://github.com/bitcoin/bitcoin/pull/25487#discussion_r917200540
  - https://github.com/bitcoin/bitcoin/pull/25487#discussion_r917203102
  - https://github.com/bitcoin/bitcoin/pull/25487#discussion_r920566511

dongcarl commented on src/mempool_args.h:
```diff
@@ -10,6 +10,9 @@ namespace kernel {
 };
 
+/** Default for -persistmempool */
+static constexpr bool DEFAULT_PERSIST_MEMPOOL{true};
```
Fixed in adf4a30d95

ryanofsky reviewed (approved):
Code review ACK 3a9dc8e675f41228cc49c3b8140c3ca39d94001f. Just some comments fixups since last review

ryanofsky commented on src/test/fuzz/validation_load_mempool.cpp:
```diff
@@ -36,9 +37,12 @@ FUZZ_TARGET_INIT(validation_load_mempool, initialize_validation_load_mempool)
         return fuzzed_file_provider.open();
     };
-    (void)LoadMempool(pool, g_setup->m_node.chainman->ActiveChainstate(), fuzzed_fopen);
+    (void)chainstate.LoadMempool(MempoolPath(g_setup->m_args), fuzzed_fopen);
```
In commit "LoadMempool: Pass in load_path, stop using gArgs" (678036771720433bd789e68e43c0e250cd3706b8)

What's the reason this fuzz test is changed to use chainstate.LoadMempool instead of kernel::LoadMempool? It seems to make test setup more fragile and LoadMempool/DumpMempool sequence a little less symmetric. Maybe extra test coverage is worth it though or there was a different reason previous code didn't work. Just curious

dongcarl replied:
Added more context in c4249c1243981f42091950d1295c2be6bd4bf118, also added a followup to the project ticket to make loadmempool less footgunny (this footgun was existed prior to this PR)

dongcarl commented:
Push 3a9dc8e675 -> 7ae032e075
- Merged `fuzz/validation_load_mempool` changes into the `DummyChainstate` commit, added git message for why calling `LoadMempool` via `CChainState` is more correct.

ryanofsky reviewed (approved):
Code review ACK 7ae032e075caa97c790603e768c7c02b80358968. Since last review commits were just rearranged internally and a fuzz change was better explained. No changes to the overall diff

glozow reviewed (commented):
light code review ACK 7ae032e075caa97c790603e768c7c02b80358968

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4726,7 +4728,9 @@ bool LoadMempool(CTxMemPool& pool, CChainState& active_chainstate, FopenFn mocka
 bool DumpMempool(const CTxMemPool& pool, FopenFn mockable_fopen_function, bool skip_file_commit)
 {
-    int64_t start = GetTimeMicros();
+    using SteadyClock = std::chrono::steady_clock;
```
9c832696826449236e88ff2fbb43d12d3d4ebc78 This can probably be added to util/time.h:


```diff
diff --git a/src/util/time.h b/src/util/time.h
index 0f87d66c2e..b01be93f83 100644
--- a/src/util/time.h
+++ b/src/util/time.h
@@ -24,6 +24,7 @@ struct NodeClock : public std::chrono::system_clock {
 };
 using NodeSeconds = std::chrono::time_point<NodeClock, std::chrono::seconds>;
 
+using SteadyClock = std::chrono::steady_clock;
 using SteadySeconds = std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>;
 using SteadyMilliseconds = std::chrono::time_point<std::chrono::steady_clock, std::chrono::milliseconds>;
 using SteadyMicroseconds = std::chrono::time_point<std::chrono::steady_clock, std::chrono::microseconds>;

MarcoFalke commented on src/init.cpp:
```diff
@@ -111,6 +112,8 @@ using node::CleanupBlockRevFiles;
 using node::DEFAULT_STOPAFTERBLOCKIMPORT;
 using node::LoadChainstate;
+using node::MempoolPath;
+using node::ShouldPersistMempool;
```
04f72e90204c049d5d320388fda1f01e53417938: What is the point of those, when it is less code overall with them removed?

dongcarl replied:
Do you mean why have these `using` declarations instead of using `node::MempoolPath`, `node::ShouldPersistMempool` directly?

MarcoFalke replied:
Yes. If only one symbol requires them, it seems less code to just write the namespace out.

dongcarl replied:
In the longer term I think `init.cpp` will go in the `node` namespace so it's probably less code change in the longer term to have the `using` declarations. In any case, probably not going to address it in this PR.

MarcoFalke commented on src/Makefile.am:
```diff
@@ -380,6 +381,7 @@ libbitcoin_node_a_SOURCES = \
   node/context.cpp \
   node/eviction.cpp \
   node/interfaces.cpp \
+  node/mempool_persist_args.cpp \
```
04f72e90204c049d5d320388fda1f01e53417938: Add to iwyu CI?

dongcarl replied:
I didn't do this most because IWYU insists that:

```
node/mempool_persist_args.cpp should add these lines:
#include <memory>         // for allocator
```

And I'm not sure why or if we need it... Does @ryanofsky know?

MarcoFalke replied:
Ok nvm then. Looks like an iwyu bug

MarcoFalke commented on src/test/fuzz/validation_load_mempool.cpp:
```diff
@@ -36,9 +37,12 @@ FUZZ_TARGET_INIT(validation_load_mempool, initialize_validation_load_mempool)
         return fuzzed_file_provider.open();
     };
-    (void)LoadMempool(pool, g_setup->m_node.chainman->ActiveChainstate(), fuzzed_fopen);
+    (void)chainstate.LoadMempool(g_setup->m_args);
```
c4249c1243981f42091950d1295c2be6bd4bf118: You are removing fuzzed_open. This will disable the fuzz tests. 


Also there is a typo in the commit message "it definitely is differece".

dongcarl replied:
Fixed in b857ac60d9

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <flatfile.h>
 #include <fs.h>
 #include <hash.h>
+#include <node/mempool_persist_args.h>
```
c4249c1243981f42091950d1295c2be6bd4bf118: Why is this needed?

dongcarl replied:
Fixed in 06b88ffb8a

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -893,6 +894,6 @@ void ThreadImport(ChainstateManager& chainman, std::vector<fs::path> vImportFile
         }
     } // End scope of CImportingNow
-    chainman.ActiveChainstate().LoadMempool(args);
+    chainman.ActiveChainstate().LoadMempool(mempool_path);
```
c4249c1243981f42091950d1295c2be6bd4bf118: I think you probably wanted to call `ShouldPersistMempool(args) ? MempoolPath(args) : fs::path{}` here? Which also makes more sense to me, as `args` is already available in this scope.

dongcarl replied:
I guess that depends on whether we want `ThreadImport` to be part of `libbitcoinkernel` or not. I'm operating under the assumption that we do want it to be part of `libbitcoinkernel`, so whenever there's an opportunity to untangle it from `ArgsManager`, I take it.

MarcoFalke commented on src/validation.h:
```diff
@@ -415,8 +415,6 @@ enum class CoinsCacheSizeState
     OK = 0
 };
 
-using FopenFn = std::function<FILE*(const fs::path&, const char*)>;
```
3298c26002a24dd196348b9849e6bc0b2e5e56fd: Instead of moving this twice in the same pull, what about only moving it only once to the final place in one go?

dongcarl replied:
Moved once in b3267258b0

maflcko reviewed (commented):
7ae032e075caa97c790603e768c7c02b80358968 📃

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

7ae032e075caa97c790603e768c7c02b80358968 📃
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgfUgwAkIj0v/a6TKMxsnCnOXt7Fx6T4Om/7AP2yIltNiFCbNSi3pK25loSjt0y
XojYPih/7ZrUggAqOANaRBg1nplwDqpiD7XjFaP9LFvXu1xkfEww1FHgXY2q25CI
V8xA69Oq2BLKfr3Zj8Lv5PQPq1dZgyaR3QmBBR7SscbSnte+SzsnOuVxmUdaCh4A
koGKy7uFEQi2IkeeFOTippHuGkLYm2jN/8TwGzmWy27KaOkjRyezbHp1ozmimB+g
/OdoV0Fve6Jzlv1ThIJeR9K8eCJhJLDC8nRecMwLehMsGOuoeJ8SozmbSihqXcFI
8VQ+WKn0WBfA0UAynC66QKPUBYthsQ5VXZ9QIS81J4Ktjkv2+CHCufyLX0mwxByy
NSu8L4zBEpojnoQTDQX35tDuWMZYWK1wQ2gBmTBBMulF6Q+VLGVecsAUXg4mGnuX
KkqC1aW6lLF//yYtcHpV/qPtBJ7Ym1yHJMR9kj5JUB8EFVaIZAnljgRjXYXy70mT
4NeqZUCC
=PbsG
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented:
Looks good, but I think there is a bug https://github.com/bitcoin/bitcoin/pull/25487#discussion_r922231747, which should be addressed or replied to before merge.

maflcko reviewed (approved):
ACK cb3e9a1e3f 🔒

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cb3e9a1e3f 🔒
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgBIgv+I0xufL+4M1xQ03SLaGjEKhoV4ts5l8hXo1Md5hP8qXMKMoBz8GOzM6vX
ctVleKfacPJQP65lTM4JqaORnTY/CA1eZz59klYy6+W5TjxasH0Tj55oVRGjA1Be
LdJpc6x9I8B9nJfvcPM5FFcBo2y7QbOKDcZQtjqi6gxdoOHsteHEts3D2CPPfuXc
UmpffGr8mQXyC6L3ISoSaMA19a4FyA8+LrhZTS2WRKvwRC55h8G53k6tIFk5HU/K
T65bW0IgVyXT1XjhF94J2EjwTX9pLWUDEeJhib7rc42onmIBKhqCEU4AWConz8G9
rGAJCkuuoPRTNwO26tmi2ZG2pBRSfGZot6AccKH57UAHgjX7vTB5Vs00IwXa84B6
cHHzR/9SYUx9af+2ciyUn2BNmKktRiVeInYjGoDONoKpuT8emF0M/QgFAelYRu6u
N9E7LKKYztGiKnxlXVxVA5AyvJGweCK2AhLU8QHA7d6jpB9ghuHFzl+7upBbgHPd
raQhDcGj
=8bBU
-----END PGP SIGNATURE-----
```


</details>

ryanofsky reviewed (approved):
Code review ACK cb3e9a1e3f8d72daaa361fc45dd853775e754b9d
 Main change is fuzzed_fopen fix, and there were some other cleanups.

glozow commented:
re ACK cb3e9a1e3f8d72daaa361fc45dd853775e754b9d via `git range-diff 7ae032e...cb3e9a1`
